### DISCUSSION TITLE
Remove dependency on typing_extensions package (Python >= 3.8)

### DISCUSSION THREAD

**Discussion by mcepl** [bokeh/bokeh#11310] [General]
Numpy doesn’t support 3.6 any more and  in case you skip dependency on 3.7 (as many Linux distributions do), objects provided by the `typing_extensions` package are not necessary on 3.8+. 

[This patch](https://github.com/bokeh/bokeh/files/6561987/stdlib-typing_extensions.patch.txt) removes this dependency.

**Comment by mattpap** [comment]
`typing_extensions` may still be useful, because we would like to take advantage of the new `TypeGuard` type (https://www.python.org/dev/peps/pep-0647/), which will be available officially in Python 3.10.

**Comment by bryevdv** [comment]
>  in case you skip dependency on 3.7 

We have no plans to do this

**Comment by mcepl** [comment]
Hmm, perhaps, this is all wrong. `typing_extensions` will never go away and it doesn’t make sense to try.

**Comment by bryevdv** [comment]
`typing_extensions` may only be removed as soon as nothing is needed from it. At bare minimum that won't be until we change to make Python 3.8 is the minimum Python version, and that will not happen in the immediate term (we only just made 3.7 the minimum for the upcoming Bokeh 2.4 version). However, if we decide to employ `TypeGuard` then it would not be until Python 3.10 is made the minimum version, and that's considerably further in the future. 

This doesn't seem like a pressing  concern, though? As dependencies go, `typing_extensions` is unproblematic to install with either conda or pip, is actively maintained by the Python devs directly, and is relatively small (~10s of kb). Is there some other specific problem or concern that weighs in on this?

[note: moving to discussion]
